Query refactoring: IdsQuery

cbuescher opened the pull request:
Split the parse(QueryParseContext ctx) method into a parsing and a query building part, adding Streamable for serialization and hashCode(), equals() for better testing.
Add basic unit test for Builder and Parser.

PR goes agains query-refacoring feature branch.

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -19,28 +19,42 @@
+    float boost = 1.0f;
 
-    private String queryName;
+    String queryName;
```
comment about visibility too, why not private with getters?

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -19,28 +19,42 @@
+public class IdsQueryBuilder extends BaseQueryBuilder implements Streamable, BoostableQueryBuilder<IdsQueryBuilder> {
 
-    private final List<String> types;
+    Collection<String> types = new ArrayList<String>();
```
can you remove generic warnings in this class please? new ArrayList<String> here should be new ArrayList<> .

dakrone replied:
Also, is there a reason to use `Collection` instead of `List<String> = new ArrayList<>();`?

cbuescher replied:
@dakrone Yes, in toQuery() types need to be compatible with result of `parseContext.queryTypes()` or `parseContext.mapperService().types()` which return `Collection<String>`. I tried avoiding re-adding or converting to list here. I could have left `values` a List here, but changed that to Collection as well for coherency.

Commit:
Query refactoring: refactored IdsQuery and added test

cbuescher commented:
I rebased this PR on current state of feature branch, added the changes to BaseQuery and the QueryBuilder interface that we already discussed in #10669.

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -108,4 +149,84 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     }
+
+    public Query toQuery(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        ArrayList<BytesRef> ids = new ArrayList<>();
```
can we init the arraylist providing its size as input since we know it? also maybe use List<BytesRef> on the left side?

cbuescher replied:
done

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -108,4 +149,84 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+        int valueSize = in.readInt();
+        for (int i = 0; i < valueSize; i++) {
+            ids.add(in.readString());
+        }
```
we don't have a utility method to read and write a Collection right? that would save us few lines of code. Shall we add it instead of repeating loops?

cbuescher replied:
I saw we have already cover Lists in Streamoutput#writeGenericValue, so I changed internal type of ids/type to List and used that. Only downside is an unchecked cast in readFrom, but that should be okay.

javanna commented:
I quickly went through it and it LGTM besides the few minor comments I left. @dakrone can you have another look too please?

cbuescher commented:
Added changes adressing the comments and rebased on current feature branch. Will wait for another look from @dakrone.

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -19,46 +19,74 @@
 
     public IdsQueryBuilder(String... types) {
-        this.types = types == null ? null : Arrays.asList(types);
+        this.types = (types == null || types.length == 0) ? new ArrayList<String>() : Arrays.asList(types);
```
you dont need the type here `new ArrayList<>()` right? your IDE should give you a warning for this type of things.

cbuescher replied:
Actually I get a compile error when I don't use the type here. Maybe because of the ternary operator here.

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -108,4 +150,73 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        this.types = (List<String>) in.readGenericValue();
+        this.ids = (List<String>) in.readGenericValue();
```
sorry but I must admit I am not too happy with using readGenericValue when we know the type upfront, loops were probably better then. Either add back loops, or add support for collections to our serialization layer, or convert to array which we already support?

cbuescher replied:
You mean re-adding the loop just for reading? read/writeGenericValue in the StreamInput/Output does support List, so except for the cast I don't understand the difference when I either use the loops as before or add a special read/write for List<String> to the Stream classes. Or maybe I don't understand.

javanna replied:
I mean both read/writeGenericValue. They are supposed to be used for an object and the method contains a lot of instanceof checks. I would love to call a method that accepts a List, thats it, cause we know upfront the type. The difference between the loops and a specialized method is less lines of code everytime you have a list that you need to serialize, the loop is going to be in the utility method. Makes sense?

cbuescher replied:
Sorry for the confusion, makes sense now.

javanna replied:
no problem at all!

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryParser.java:
```diff
@@ -107,26 +98,13 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        query.setBoost(boost);
-        if (queryName != null) {
-            parseContext.addNamedQuery(queryName, query);
+            throw new QueryParsingException(parseContext, "[ids] query, no ids values provided");
```
this check should probably be moved to the validate method? otherwise ids can potentially be empty when the request gets created via java api (doesn't go through fromXContent). Makes sense?

cbuescher replied:
This one is weird in the code before the refactoring:
```
if (!idsProvided) {
   throw new QueryParsingException(parseContext, "[ids] query, no ids values provided");
}
if (ids.isEmpty()) {
    return Queries.newMatchNoDocsQuery();
}
```
I wonder if this distiction makes sense, seems to be okay at the moment to provide no ids as longs as "values" field is set in the query, even if its empty.

cbuescher replied:
To keep the current behaviour, I think we should allow no ids, since this is currently not throwing an error. However, the current parser checks for existence of "values", I wonder if that can go?

javanna replied:
> I wonder if this distiction makes sense

testing is the way to find out, makes no sense to me either...

> I think we should allow no ids

not sure to be honest, What I care about is aligning validation between rest layer and java api. If we accept empty list let's accept it everywhere.

cbuescher replied:
Talked to clint about this, the difference here is empty ids list, which is okay and should return no docs, and missing "value" field in query dsl, which throws an exception, but this is only affecting parsing the json. So I would opt for continue throwing the parsing exception, but validate() which operated on the query object should allow empty ids list.

javanna replied:
ok sounds good

javanna commented on src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -108,4 +150,73 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+
+        List<BytesRef> ids = new ArrayList<>(this.ids.size());
+        for (String value : this.ids) {
+            BytesRef ref = new BytesRef(value);
```
wondering, should we use BytesRef from the beginning rather than string?

cbuescher replied:
I just realized that Uid.createTypeUids later on converts any List<? extends Object) to BytesRef anyway, so will delete this block. I'd like to String ids in the query class for better readability, logging etc.

javanna commented:
I just had another look, left a few comments

cbuescher committed:
Removed unnecessary conversion to BytesRef

cbuescher committed:
Added test for check for QueryParsingException when value field is not set

cbuescher committed:
Add writing and reading of String lists to StreamIO classes

cbuescher commented:
Went over the comments and added serialization of string lists to the StreamInput/Output classes, also added test for checking exception on missing value field.

javanna commented on src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTest.java:
```diff
@@ -0,0 +1,91 @@
+     * @throws IOException
+     */
+    @Test(expected=QueryParsingException.class)
+    public void idsNotProvided() throws IOException {
```
can you prepend the name of the method with test, just to make sure?

javanna commented on src/test/java/org/elasticsearch/index/query/IdsQueryBuilderTest.java:
```diff
@@ -0,0 +1,91 @@
+            assertThat(query, is(instanceOf(TermsQuery.class)));
+            TermsQuery termQuery = (TermsQuery) query;
+            assertThat(termQuery.getBoost(), is(testQuery.boost()));
+            // because internals of TermsFilter are well hidden, check string representation
```
s/TermsFilter/TermsQuery ?

javanna commented on src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -335,6 +335,21 @@ public final Boolean readOptionalBoolean() throws IOException {
+            ret.add(readString());
+        }
+        return ret;
+    }
```
should we make this change upstream?

cbuescher replied:
Yes, separate PR or can I just push this?

javanna replied:
I'd say separate PR against master

cbuescher replied:
Opened https://github.com/elastic/elasticsearch/pull/11056 for this.

cbuescher committed:
Add test prefix to testmethod, minor change in comment

cbuescher commented:
Just rebased this on current state of feature branch which now includes the addition if streaming string lists & adressed the last two comments.

javanna commented:
LGTM
